File: src/go/4_BehavioralPatterns/8_State/handmade.go
package state

import (
  "bufio"
  "fmt"
  "os"
  "strconv"
)

type State int

const (
  OffHook State = iota
  Connecting
  Connected
  OnHold
  OnHook
)

func (s State) String() string {
  switch s {
  case OffHook: return "OffHook"
  case Connecting: return "Connecting"
  case Connected: return "Connected"
  case OnHold: return "OnHold"
  case OnHook: return "OnHook"
  }
  return "Unknown"
}

type Trigger int

const (
  CallDialed Trigger = iota
  HungUp
  CallConnected
  PlacedOnHold
  TakenOffHold
  LeftMessage
)

func (t Trigger) String() string {
  switch t {
  case CallDialed: return "CallDialed"
  case HungUp: return "HungUp"
  case CallConnected: return "CallConnected"
  case PlacedOnHold: return "PlacedOnHold"
  case TakenOffHold: return "TakenOffHold"
  case LeftMessage: return "LeftMessage"
  }
  return "Unknown"
}

type TriggerResult struct {
  Trigger Trigger
  State State
}

var rules = map[State][]TriggerResult {
  OffHook: {
    {CallDialed, Connecting},
  },
  Connecting: {
    {HungUp, OffHook},
    {CallConnected, Connected},
  },
  Connected: {
    {LeftMessage, OnHook},
    {HungUp, OnHook},
    {PlacedOnHold, OnHold},
  },
  OnHold: {
    {TakenOffHold, Connected},
    {HungUp, OnHook},
  },
}

func main() {
  state, exitState := OffHook, OnHook
  for ok := true; ok; ok = state != exitState {
    fmt.Println("The phone is currently", state)
    fmt.Println("Select a trigger:")

    for i := 0; i < len(rules[state]); i++ {
      tr := rules[state][i]
      fmt.Println(strconv.Itoa(i), ".", tr.Trigger)
    }

    input, _, _ := bufio.NewReader(os.Stdin).ReadLine()
    i, _ := strconv.Atoi(string(input))

    tr := rules[state][i]
    state = tr.State
  }
  fmt.Println("We are done using the phone")
}

File: src/js/4_BehavioralPatterns/8_State/handmade.js
const readline = require('readline');

let rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout
});

let State = Object.freeze({
  offHook: 'off hook',
  connecting: 'connecting',
  connected: 'connected',
  onHold: 'on hold',
  onHook: 'on hook'
});

let Trigger = Object.freeze({
  callDialed: 'dial a number',
  hungUp: 'hang up',
  callConnected: 'call is connected',
  placedOnHold: 'placed on hold',
  takenOffHold: 'taken off hold',
  leftMessage: 'leave a message'
});

let rules = {};
rules[State.offHook] = [
  {
    trigger: Trigger.callDialed,
    state: State.connecting
  }
];
rules[State.connecting] = [
  {
    trigger: Trigger.hungUp,
    state: State.onHook
  },
  {
    trigger: Trigger.callConnected,
    state: State.connected
  }
];
rules[State.connected] = [
  {
    trigger: Trigger.leftMessage,
    state: State.onHook
  },
  {
    trigger: Trigger.hungUp,
    state: State.onHook
  },
  {
    trigger: Trigger.placedOnHold,
    state: State.onHold
  }
];
rules[State.onHold] = [
  {
    trigger: Trigger.takenOffHold,
    state: State.connected
  },
  {
    trigger: Trigger.hungUp,
    state: State.onHook
  }
];

let state = State.offHook;
let exitState = State.onHook;

let getInput = function() {
  let prompt = [
    `The phone is currently ${state}`,
    'What\'s next:'
  ];

  for (let i = 0; i < rules[state].length; ++i)
  {
    let t = rules[state][i].trigger;
    prompt.push(`${i}. ${t}`);
  }

  // force an extra line break
  prompt.push('');

  rl.question(prompt.join('\n'), function(answer)
  {
    let input = parseInt(answer);
    state = rules[state][input].state;

    if (state !== exitState)
      getInput();
    else {
      console.log('We are done using the phone.');
      rl.close();
    }
  });
};

getInput();